lib/paseto/token.rb

```diff
@@ -0,0 +1,30 @@
+module Paseto
+  Token = Struct.new(:header, :payload, :footer) do
+    def to_message
+      message = [
+        header,
+        Paseto.encode64(payload)
+      ]
+
+      message << Paseto.encode64(footer) if footer
+
+      message.join('.')
+    end
+  end
+
+  def self.parse_raw_token(raw, expected_header = nil, expected_footer = nil)
+    version, purpose, payload, footer = raw.split('.')
+
+    header = "#{version}.#{purpose}"
+    if !expected_header.nil? && header != expected_header
```

Comment:
In what scenario would the expected_header be `nil`, should it always been defined?

Reply:
Missed this earlier, but `expected_header` is `nil` as part of `read_unauthenticated_footer`. I actually think I have a better way of doing this coming soon.